casper/src/test/scala/coop/rchain/casper/util/rholang/InterpreterUtilTest.scala

```diff
@@ -0,0 +1,154 @@
+
+  type StateWithChain[A] = State[BlockDag, A]
+  val initState   = BlockDag().copy(currentId = -1)
+  val storageSize = 1024L * 1024
+
+  "computeBlockCheckpoint" should "compute the final post-state of a chain properly" in {
+    val storageDirectory = Files.createTempDirectory("casper-interp-util-test")
+
+    val genesisDeploys = Vector(
+      "@1!(1)",
+      "@2!(2)",
+      "for(@a <- @1){ @123!(5 * a) }"
+    ).flatMap(mkTerm(_).toOption).map(ProtoUtil.termDeploy)
+
+    val b1Deploys = Vector(
+      "@1!(1)",
+      "for(@a <- @2){ @456!(5 * a) }"
+    ).flatMap(mkTerm(_).toOption).map(ProtoUtil.termDeploy)
+
+    val b2Deploys = Vector(
+      "for(@a <- @123; @b <- @456){ @1!(a + b) }"
+    ).flatMap(mkTerm(_).toOption).map(ProtoUtil.termDeploy)
+
+    val b3Deploys = Vector(
+      "@7!(7)"
+    ).flatMap(mkTerm(_).toOption).map(ProtoUtil.termDeploy)
+    /*
+     * DAG Looks like this:
+     *
+     *          b3
+     *           |
+     *          b2
+     *           |
+     *          b1
+     *           |
+     *         genesis
+     */
+    def createChain[F[_]: Monad: BlockDagState]: F[BlockMessage] =
+      for {
+        genesis <- createBlock[F](Seq.empty, deploys = genesisDeploys)
+        b1      <- createBlock[F](Seq(genesis.blockHash), deploys = b1Deploys)
+        b2      <- createBlock[F](Seq(b1.blockHash), deploys = b2Deploys)
+        b3      <- createBlock[F](Seq(b2.blockHash), deploys = b3Deploys)
+      } yield b3
+    val chain   = createChain[StateWithChain].runS(initState).value
+    val genesis = chain.idToBlocks(0)
+
+    val b1 = chain.idToBlocks(1)
+    val b3 = chain.idToBlocks(3)
+
+    val (genCh, genMap) = computeBlockCheckpoint(genesis,
+                                                 genesis,
+                                                 chain,
+                                                 storageDirectory,
+                                                 storageSize,
+                                                 Map.empty[ByteString, Checkpoint])
+    val genPostState = genCh.toTuplespace.storageRepr
+    genPostState.contains("@{2}!(2)") should be(true)
+    genPostState.contains("@{123}!(5)") should be(true)
+
+    val (b1Ch, b1Map) =
+      computeBlockCheckpoint(b1, genesis, chain, storageDirectory, storageSize, genMap)
+    val b1PostState = b1Ch.toTuplespace.storageRepr
+    b1PostState.contains("@{1}!(1)") should be(true)
+    b1PostState.contains("@{123}!(5)") should be(true)
+    b1PostState.contains("@{456}!(10)") should be(true)
+
+    //note skipping of b2 to force a test of the recursive aspect of computeBlockCheckpoint
+
+    val (b3Ch, _) =
+      computeBlockCheckpoint(b3, genesis, chain, storageDirectory, storageSize, b1Map)
+    val b3PostState = b3Ch.toTuplespace.storageRepr
+    b3PostState.contains("@{1}!(1)") should be(true)
+    b3PostState.contains("@{1}!(15)") should be(true)
+    b3PostState.contains("@{7}!(7)") should be(true)
```

Comment:
You could probably factor out the assertions to the method that will take into the block, genesis, chain, storage, …, and deploys the post state should contain.